File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/PhasorValueBase.vb
'*******************************************************************************************************
'  PhasorValueBase.vb - Phasor value base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Note: Phasors are stored in rectangular format internally
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports Tva.Math

' This class represents the protocol independent representation of a phasor value.
<CLSCompliant(False)> _
Public MustInherit Class PhasorValueBase

    Inherits ChannelValueBase(Of IPhasorDefinition)
    Implements IPhasorValue

    Protected Delegate Function CreateNewPhasorValueFunctionSignature(ByVal parent As IDataCell, ByVal phasorDefintion As IPhasorDefinition, ByVal real As Single, ByVal imaginary As Single) As IPhasorValue

    Private m_real As Single
    Private m_imaginary As Single
    Private m_compositeValues As CompositeValues

    ' Create phasor from polar coordinates (angle expected in Degrees)
    ' Note: This method is expected to be implemented as a public shared method in derived class automatically passing in createNewPhasorValueFunction
    Protected Shared Function CreateFromPolarValues(ByVal createNewPhasorValueFunction As CreateNewPhasorValueFunctionSignature, ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal angle As Single, ByVal magnitude As Single) As IPhasorValue

        Return CreateFromRectangularValues(createNewPhasorValueFunction, parent, phasorDefinition, CalculateRealComponent(angle, magnitude), CalculateImaginaryComponent(angle, magnitude))

    End Function

    ' Create phasor from rectangular coordinates
    ' Note: This method is expected to be implemented as a public shared method in derived class automatically passing in createNewPhasorValueFunction
    Protected Shared Function CreateFromRectangularValues(ByVal createNewPhasorValueFunction As CreateNewPhasorValueFunctionSignature, ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal real As Single, ByVal imaginary As Single) As IPhasorValue

        If phasorDefinition Is Nothing Then Throw New ArgumentNullException("No phasor definition specified")
        Return createNewPhasorValueFunction.Invoke(parent, phasorDefinition, real, imaginary)

    End Function

    ' Create phasor from unscaled rectangular coordinates
    ' Note: This method is expected to be implemented as a public shared method in derived class automatically passing in createNewPhasorValueFunction
    Protected Shared Function CreateFromUnscaledRectangularValues(ByVal createNewPhasorValueFunction As CreateNewPhasorValueFunctionSignature, ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal real As Int16, ByVal imaginary As Int16) As IPhasorValue

        Dim factor As Single = phasorDefinition.ConversionFactor
        Return CreateFromRectangularValues(createNewPhasorValueFunction, parent, phasorDefinition, real * factor, imaginary * factor)

    End Function

    ' Gets real component from angle (in Degrees) and magnitude
    Public Shared Function CalculateRealComponent(ByVal angle As Single, ByVal magnitude As Single) As Single

        Return magnitude * System.Math.Cos(angle * System.Math.PI / 180)

    End Function

    ' Gets imaginary component from angle (in Degrees) and magnitude
    Public Shared Function CalculateImaginaryComponent(ByVal angle As Single, ByVal magnitude As Single) As Single

        Return magnitude * System.Math.Sin(angle * System.Math.PI / 180)

    End Function

    ' Calculate watts from imaginary and real components of two phasors
    Public Shared Function CalculatePower(ByVal voltage As IPhasorValue, ByVal current As IPhasorValue) As Single

        If voltage Is Nothing Then Throw New ArgumentNullException("No voltage specified")
        If current Is Nothing Then Throw New ArgumentNullException("No current specified")

        Return 3 * (voltage.Real * current.Real + voltage.Imaginary * current.Imaginary)
        'Return 3 * voltage.Magnitude * current.Magnitude * System.Math.Cos((voltage.Angle - current.Angle) * System.Math.PI / 180)

    End Function

    ' Calculate vars from imaginary and real components of two phasors
    Public Shared Function CalculateVars(ByVal voltage As IPhasorValue, ByVal current As IPhasorValue) As Single

        If voltage Is Nothing Then Throw New ArgumentNullException("No voltage specified")
        If current Is Nothing Then Throw New ArgumentNullException("No current specified")

        Return 3 * (voltage.Imaginary * current.Real - voltage.Real * current.Imaginary)
        'Return 3 * voltage.Magnitude * current.Magnitude * System.Math.Sin((voltage.Angle - current.Angle) * System.Math.PI / 180)

    End Function

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize phasor value
        m_real = info.GetSingle("real")
        m_imaginary = info.GetSingle("imaginary")

    End Sub

    Protected Sub New(ByVal parent As IDataCell)

        MyBase.New(parent)

        m_compositeValues = New CompositeValues(2)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal real As Single, ByVal imaginary As Single)
    Protected Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal real As Single, ByVal imaginary As Single)

        MyBase.New(parent, phasorDefinition)

        m_real = real
        m_imaginary = imaginary
        m_compositeValues = New CompositeValues(2)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal unscaledReal As Int16, ByVal unscaledImaginary As Int16)
    Protected Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal unscaledReal As Int16, ByVal unscaledImaginary As Int16)

        MyClass.New(parent, phasorDefinition, unscaledReal * phasorDefinition.ConversionFactor, unscaledImaginary * phasorDefinition.ConversionFactor)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    Protected Sub New(ByVal parent As IDataCell, ByVal phasorDefinition As IPhasorDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, phasorDefinition)
        ParseBinaryImage(Nothing, binaryImage, startIndex)
        m_compositeValues = New CompositeValues(2)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal phasorValue As IPhasorValue)
    Protected Sub New(ByVal phasorValue As IPhasorValue)

        MyClass.New(phasorValue.Parent, phasorValue.Definition, phasorValue.Real, phasorValue.Imaginary)

    End Sub

    Public Overridable ReadOnly Property CoordinateFormat() As CoordinateFormat Implements IPhasorValue.CoordinateFormat
        Get
            Return Definition.CoordinateFormat
        End Get
    End Property

    Public Overridable ReadOnly Property Type() As PhasorType Implements IPhasorValue.Type
        Get
            Return Definition.Type
        End Get
    End Property

    Public Overridable Property Angle() As Single Implements IPhasorValue.Angle
        Get
            Return System.Math.Atan2(m_imaginary, m_real) * 180 / System.Math.PI
        End Get
        Set(ByVal value As Single)
            ' We store angle as one of our required composite values
            m_compositeValues(CompositePhasorValue.Angle) = value

            ' If all composite values have been received, we can calculate phasor's real and imaginary values
            CalculatePhasorValueFromComposites()
        End Set
    End Property

    Public Overridable ReadOnly Property AngleReceived() As Boolean
        Get
            Return m_compositeValues.Received(CompositePhasorValue.Angle)
        End Get
    End Property

    Public Overridable Property Magnitude() As Single Implements IPhasorValue.Magnitude
        Get
            Return System.Math.Sqrt(m_real * m_real + m_imaginary * m_imaginary)
        End Get
        Set(ByVal value As Single)
            m_compositeValues(CompositePhasorValue.Magnitude) = value

            ' If all composite values have been received, we can calculate phasor's real and imaginary values
            CalculatePhasorValueFromComposites()
        End Set
    End Property

    Public Overridable ReadOnly Property MagnitudeReceived() As Boolean
        Get
            Return m_compositeValues.Received(CompositePhasorValue.Magnitude)
        End Get
    End Property

    Private Sub CalculatePhasorValueFromComposites()

        If m_compositeValues.AllReceived Then
            Dim angle, magnitude As Single

            ' All values received, create a new phasor value from composite values
            angle = m_compositeValues(CompositePhasorValue.Angle)
            magnitude = m_compositeValues(CompositePhasorValue.Magnitude)

            m_real = CalculateRealComponent(angle, magnitude)
            m_imaginary = CalculateImaginaryComponent(angle, magnitude)
        End If

    End Sub

    Public Overridable Property Real() As Single Implements IPhasorValue.Real
        Get
            Return m_real
        End Get
        Set(ByVal value As Single)
            m_real = value
        End Set
    End Property

    Public Overridable Property Imaginary() As Single Implements IPhasorValue.Imaginary
        Get
            Return m_imaginary
        End Get
        Set(ByVal value As Single)
            m_imaginary = value
        End Set
    End Property

    Public Overridable Property UnscaledReal() As Int16 Implements IPhasorValue.UnscaledReal
        Get
            Return Convert.ToInt16(m_real / Definition.ConversionFactor)
        End Get
        Set(ByVal value As Int16)
            m_real = value * Definition.ConversionFactor
        End Set
    End Property

    Public Overridable Property UnscaledImaginary() As Int16 Implements IPhasorValue.UnscaledImaginary
        Get
            Return Convert.ToInt16(m_imaginary / Definition.ConversionFactor)
        End Get
        Set(ByVal value As Int16)
            m_imaginary = value * Definition.ConversionFactor
        End Set
    End Property

    Default Public Overrides Property CompositeValue(ByVal index As Integer) As Single
        Get
            Select Case index
                Case CompositePhasorValue.Angle
                    Return Angle
                Case CompositePhasorValue.Magnitude
                    Return Magnitude
                Case Else
                    Throw New IndexOutOfRangeException("Specified phasor value composite index, " & index & ", is out of range - there are only two composite values for a phasor value: angle (0) and magnitude (1)")
            End Select
        End Get
        Set(ByVal value As Single)
            Select Case index
                Case CompositePhasorValue.Angle
                    Angle = value
                Case CompositePhasorValue.Magnitude
                    Magnitude = value
                Case Else
                    Throw New IndexOutOfRangeException("Specified phasor value composite index, " & index & ", is out of range - there are only two composite values for a phasor value: angle (0) and magnitude (1)")
            End Select
        End Set
    End Property

    Public Overrides ReadOnly Property CompositeValueCount() As Integer
        Get
            Return 2
        End Get
    End Property

    Public Overrides ReadOnly Property IsEmpty() As Boolean
        Get
            Return (m_real = 0 And m_imaginary = 0)
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            If DataFormat = Phasors.DataFormat.FixedInteger Then
                Return 4
            Else
                Return 8
            End If
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = CreateArray(Of Byte)(BodyLength)

            If CoordinateFormat = Phasors.CoordinateFormat.Rectangular Then
                If DataFormat = Phasors.DataFormat.FixedInteger Then
                    EndianOrder.BigEndian.CopyBytes(UnscaledReal, buffer, 0)
                    EndianOrder.BigEndian.CopyBytes(UnscaledImaginary, buffer, 2)
                Else
                    EndianOrder.BigEndian.CopyBytes(m_real, buffer, 0)
                    EndianOrder.BigEndian.CopyBytes(m_imaginary, buffer, 4)
                End If
            Else
                If DataFormat = Phasors.DataFormat.FixedInteger Then
                    EndianOrder.BigEndian.CopyBytes(Convert.ToUInt16(Magnitude), buffer, 0)
                    EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(Angle * System.Math.PI / 180 * 10000), buffer, 2)
                Else
                    EndianOrder.BigEndian.CopyBytes(Magnitude, buffer, 0)
                    EndianOrder.BigEndian.CopyBytes(Convert.ToSingle(Angle * System.Math.PI / 180), buffer, 4)
                End If
            End If

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Integer)

        If CoordinateFormat = Phasors.CoordinateFormat.Rectangular Then
            If DataFormat = Phasors.DataFormat.FixedInteger Then
                UnscaledReal = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
                UnscaledImaginary = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2)
            Else
                m_real = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex)
                m_imaginary = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex + 4)
            End If
        Else
            Dim magnitude As Single
            Dim angle As Single

            If DataFormat = Phasors.DataFormat.FixedInteger Then
                magnitude = Convert.ToSingle(EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex))
                angle = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2) * 180 / System.Math.PI / 10000
            Else
                magnitude = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex)
                angle = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex + 4) * 180 / System.Math.PI
            End If

            m_real = CalculateRealComponent(angle, magnitude)
            m_imaginary = CalculateImaginaryComponent(angle, magnitude)
        End If

    End Sub

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize phasor value
        info.AddValue("real", m_real)
        info.AddValue("imaginary", m_imaginary)

    End Sub

End Class

File: Source/Libraries/GSF.PhasorProtocols/PhasorValueBase.cs
//******************************************************************************************************
//  PhasorValueBase.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  11/12/2004 - J. Ritchie Carroll
//       Generated original version of source code.
//  12/17/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using GSF.TimeSeries;
using GSF.Units;
using GSF.Units.EE;

namespace GSF.PhasorProtocols
{
    #region [ Enumerations ]

    /// <summary>
    /// Composite polar value indices enumeration.
    /// </summary>
    public enum CompositePhasorValue
    {
        /// <summary>
        /// Composite angle value index.
        /// </summary>
        Angle,
        /// <summary>
        /// Composite magnitude value index.
        /// </summary>
        Magnitude
    }

    #endregion

    /// <summary>
    /// Represents the common implementation of the protocol independent representation of a phasor value.
    /// </summary>
    [Serializable]
    public abstract class PhasorValueBase : ChannelValueBase<IPhasorDefinition>, IPhasorValue
    {
        #region [ Members ]

        // Fields
        private ComplexNumber m_phasor;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="PhasorValueBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="PhasorValueBase"/>.</param>
        /// <param name="phasorDefinition">The <see cref="IPhasorDefinition"/> associated with this <see cref="PhasorValueBase"/>.</param>
        protected PhasorValueBase(IDataCell parent, IPhasorDefinition phasorDefinition)
            : base(parent, phasorDefinition)
        {
        }

        /// <summary>
        /// Creates a new <see cref="PhasorValueBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="PhasorValueBase"/>.</param>
        /// <param name="phasorDefinition">The <see cref="IPhasorDefinition"/> associated with this <see cref="PhasorValueBase"/>.</param>
        /// <param name="real">The real value of this <see cref="PhasorValueBase"/>.</param>
        /// <param name="imaginary">The imaginary value of this <see cref="PhasorValueBase"/>.</param>
        protected PhasorValueBase(IDataCell parent, IPhasorDefinition phasorDefinition, double real, double imaginary)
            : base(parent, phasorDefinition)
        {
            m_phasor.Real = real;
            m_phasor.Imaginary = imaginary;
        }

        /// <summary>
        /// Creates a new <see cref="PhasorValueBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="PhasorValueBase"/>.</param>
        /// <param name="phasorDefinition">The <see cref="IPhasorDefinition"/> associated with this <see cref="PhasorValueBase"/>.</param>
        /// <param name="angle">The <see cref="GSF.Units.Angle"/> value (a.k.a., the argument) of this <see cref="PhasorValueBase"/>, in radians.</param>
        /// <param name="magnitude">The magnitude value (a.k.a., the absolute value or modulus) of this <see cref="PhasorValueBase"/>.</param>
        protected PhasorValueBase(IDataCell parent, IPhasorDefinition phasorDefinition, Angle angle, double magnitude)
            : base(parent, phasorDefinition)
        {
            m_phasor.Angle = angle;
            m_phasor.Magnitude = magnitude;
        }

        /// <summary>
        /// Creates a new <see cref="PhasorValueBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected PhasorValueBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize phasor value
            m_phasor.Real = info.GetDouble("real");
            m_phasor.Imaginary = info.GetDouble("imaginary");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="GSF.PhasorProtocols.CoordinateFormat"/> of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual CoordinateFormat CoordinateFormat => Definition.CoordinateFormat;

        /// <summary>
        /// Gets the <see cref="GSF.PhasorProtocols.AngleFormat"/> of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual AngleFormat AngleFormat => Definition.AngleFormat;

        /// <summary>
        /// Gets the <see cref="PhasorType"/> of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual PhasorType Type => Definition.PhasorType;

        /// <summary>
        /// Gets <see cref="GSF.Units.EE.Phasor"/> value from this <see cref="IPhasorValue"/>.
        /// </summary>
        public Phasor Phasor => new Phasor(Type, m_phasor);

        /// <summary>
        /// Gets or sets the <see cref="GSF.Units.Angle"/> value (a.k.a., the argument) of this <see cref="PhasorValueBase"/>, in radians.
        /// </summary>
        public virtual Angle Angle
        {
            get
            {
                return m_phasor.Angle;
            }
            set
            {
                m_phasor.Angle = value;
            }
        }

        /// <summary>
        /// Gets or sets the magnitude value (a.k.a., the absolute value or modulus) of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual double Magnitude
        {
            get
            {
                return m_phasor.Magnitude;
            }
            set
            {
                m_phasor.Magnitude = value;
            }
        }

        /// <summary>
        /// Gets or sets the real value of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual double Real
        {
            get
            {
                return m_phasor.Real;
            }
            set
            {
                m_phasor.Real = value;
            }
        }

        /// <summary>
        /// Gets or sets the imaginary value of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual double Imaginary
        {
            get
            {
                return m_phasor.Imaginary;
            }
            set
            {
                m_phasor.Imaginary = value;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of the real value of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual int UnscaledReal
        {
            get
            {
                unchecked
                {
                    return (int)(m_phasor.Real / Definition.ConversionFactor);
                }
            }
            set
            {
                m_phasor.Real = value * Definition.ConversionFactor;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of the imaginary value of this <see cref="PhasorValueBase"/>.
        /// </summary>
        public virtual int UnscaledImaginary
        {
            get
            {
                unchecked
                {
                    return (int)(m_phasor.Imaginary / Definition.ConversionFactor);
                }
            }
            set
            {
                m_phasor.Imaginary = value * Definition.ConversionFactor;
            }
        }

        /// <summary>
        /// Gets total number of composite values that this <see cref="PhasorValueBase"/> provides.
        /// </summary>
        public override int CompositeValueCount => 2;

        /// <summary>
        /// Gets boolean value that determines if none of the composite values of <see cref="PhasorValueBase"/> have been assigned a value.
        /// </summary>
        public override bool IsEmpty => m_phasor.NoneAssigned;

        /// <summary>
        /// Gets the length of the <see cref="BodyImage"/>.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int BodyLength => DataFormat == DataFormat.FixedInteger ? 4 : 8;

        /// <summary>
        /// Gets the binary body image of the <see cref="PhasorValueBase"/> object.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];

                // Had to make a decision on usage versus typical protocol implementation when
                // exposing values as double / int when protocols typically use float / short for
                // transmission. Exposing values as double / int makes class more versatile by
                // allowing future protocol implementations to support higher resolution values
                // simply by overriding BodyLength, BodyImage and ParseBodyImage. However, exposing
                // the double / int values runs the risk of providing values that are outside the
                // data type limitations, hence the unchecked section below. Risk should generally
                // be low in typical usage scenarios since values being transmitted via a generated
                // image were likely parsed previously from a binary image with the same constraints.
                unchecked
                {
                    if (CoordinateFormat == CoordinateFormat.Rectangular)
                    {
                        if (DataFormat == DataFormat.FixedInteger)
                        {
                            BigEndian.CopyBytes((short)UnscaledReal, buffer, 0);
                            BigEndian.CopyBytes((short)UnscaledImaginary, buffer, 2);
                        }
                        else
                        {
                            BigEndian.CopyBytes((float)m_phasor.Real, buffer, 0);
                            BigEndian.CopyBytes((float)m_phasor.Imaginary, buffer, 4);
                        }
                    }
                    else
                    {
                        if (DataFormat == DataFormat.FixedInteger)
                        {
                            BigEndian.CopyBytes((ushort)(m_phasor.Magnitude / Definition.ConversionFactor), buffer, 0);

                            if (AngleFormat == AngleFormat.Radians)
                                BigEndian.CopyBytes((short)(m_phasor.Angle * 10000.0D), buffer, 2);
                            else
                                BigEndian.CopyBytes((short)(m_phasor.Angle.ToDegrees() * 10000.0D), buffer, 2);
                        }
                        else
                        {
                            BigEndian.CopyBytes((float)m_phasor.Magnitude, buffer, 0);

                            if (AngleFormat == AngleFormat.Radians)
                                BigEndian.CopyBytes((float)m_phasor.Angle, buffer, 4);
                            else
                                BigEndian.CopyBytes((float)m_phasor.Angle.ToDegrees(), buffer, 4);
                        }
                    }
                }

                return buffer;
            }
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="PhasorValueBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Phasor Type", (int)Type + ": " + Type);
                baseAttributes.Add("Angle Value", Angle.ToDegrees() + "°");
                baseAttributes.Add("Magnitude Value", Magnitude.ToString());
                baseAttributes.Add("Real Value", Real.ToString());
                baseAttributes.Add("Imaginary Value", Imaginary.ToString());
                baseAttributes.Add("Unscaled Real Value", UnscaledReal.ToString());
                baseAttributes.Add("Unscaled Imaginary Value", UnscaledImaginary.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Gets the specified composite value of this <see cref="PhasorValueBase"/>.
        /// </summary>
        /// <param name="index">Index of composite value to retrieve.</param>
        /// <remarks>
        /// Some <see cref="ChannelValueBase{T}"/> implementations can contain more than one value, this method is used to abstractly expose each value.
        /// </remarks>
        /// <returns>A <see cref="double"/> representing the composite value.</returns>
        public override double GetCompositeValue(int index)
        {
            switch (index)
            {
                case (int)CompositePhasorValue.Angle:
                    return Angle.ToDegrees();
                case (int)CompositePhasorValue.Magnitude:
                    return Magnitude;
                default:
                    throw new ArgumentOutOfRangeException(nameof(index), "Invalid composite index requested");
            }
        }

        /// <summary>
        /// Creates a new <see cref="IMeasurement"/> value for specified composite value for this <see cref="PhasorValueBase"/>.
        /// </summary>
        /// <param name="valueIndex">Composite value index for which to derive new <see cref="IMeasurement"/> value.</param>
        /// <returns>New <see cref="IMeasurement"/> value for specified composite value for this <see cref="PhasorValueBase"/>.</returns>
        protected override IMeasurement CreateMeasurement(int valueIndex)
        {
            IMeasurement measurement = base.CreateMeasurement(valueIndex);

            if (valueIndex == (int)CompositePhasorValue.Angle)
            {
                double adjustedValue = measurement.AdjustedValue;

                if (adjustedValue <= -180 || adjustedValue > 180)
                {
                    adjustedValue = Angle.FromDegrees(adjustedValue).ToRange(-Math.PI, false).ToDegrees();
                    measurement.Value = (adjustedValue - measurement.Adder) / measurement.Multiplier;
                }
            }

            return measurement;
        }

        /// <summary>
        /// Parses the binary body image.
        /// </summary>
        /// <param name="buffer">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="buffer"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="buffer"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int ParseBodyImage(byte[] buffer, int startIndex, int length)
        {
            // Length is validated at a frame level well in advance so that low level parsing routines do not have
            // to re-validate that enough length is available to parse needed information as an optimization...

            if (DataFormat == DataFormat.FixedInteger)
            {
                if (CoordinateFormat == CoordinateFormat.Rectangular)
                {
                    // Parse from fixed-integer, rectangular
                    UnscaledReal = BigEndian.ToInt16(buffer, startIndex);
                    UnscaledImaginary = BigEndian.ToInt16(buffer, startIndex + 2);
                }
                else
                {
                    // Parse from fixed-integer, polar
                    m_phasor.Magnitude = BigEndian.ToUInt16(buffer, startIndex) * Definition.ConversionFactor;

                    if (AngleFormat == AngleFormat.Radians)
                        m_phasor.Angle = BigEndian.ToInt16(buffer, startIndex + 2) / 10000.0D;
                    else
                        m_phasor.Angle = Angle.FromDegrees(BigEndian.ToInt16(buffer, startIndex + 2) / 10000.0D);
                }

                return 4;
            }

            if (CoordinateFormat == CoordinateFormat.Rectangular)
            {
                // Parse from single-precision floating-point, rectangular
                m_phasor.Real = BigEndian.ToSingle(buffer, startIndex);
                m_phasor.Imaginary = BigEndian.ToSingle(buffer, startIndex + 4);
            }
            else
            {
                // Parse from single-precision floating-point, polar
                m_phasor.Magnitude = BigEndian.ToSingle(buffer, startIndex);

                if (AngleFormat == AngleFormat.Radians)
                    m_phasor.Angle = BigEndian.ToSingle(buffer, startIndex + 4);
                else
                    m_phasor.Angle = Angle.FromDegrees(BigEndian.ToSingle(buffer, startIndex + 4));
            }

            return 8;
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize phasor value
            info.AddValue("real", m_phasor.Real);
            info.AddValue("imaginary", m_phasor.Imaginary);
        }

        #endregion

        #region [ Static ]

        // Static Methods

        /// <summary>
        /// Calculates watts from imaginary and real components of a voltage and current phasor.
        /// </summary>
        /// <param name="voltage">Voltage phasor.</param>
        /// <param name="current">Current phasor.</param>
        /// <exception cref="ArgumentNullException"><paramref name="voltage"/> and <paramref name="current"/> must not be null.</exception>
        /// <returns>Calculated watts from imaginary and real components of specified <paramref name="voltage"/> and <paramref name="current"/> phasors.</returns>
        public static Power CalculatePower(IPhasorValue voltage, IPhasorValue current)
        {
            if (voltage == null)
                throw new ArgumentNullException(nameof(voltage), "No voltage specified");

            if (current == null)
                throw new ArgumentNullException(nameof(current), "No current specified");

            return Phasor.CalculateActivePower(voltage.Phasor, current.Phasor);
        }

        /// <summary>
        /// Calculates vars (total volt-amperes of reactive power) from imaginary and real components of a voltage and current phasor.
        /// </summary>
        /// <param name="voltage">Voltage phasor.</param>
        /// <param name="current">Current phasor.</param>
        /// <exception cref="ArgumentNullException"><paramref name="voltage"/> and <paramref name="current"/> must not be null.</exception>
        /// <remarks>
        /// Although the <see cref="Power"/> units class technically represents watts (i.e., real power) and vars (i.e., imaginary power)
        /// are properly expressed in volt-amperes reactive (VAr), the calculated result is still a representation of power and therefore
        /// the <see cref="Power"/> units class is used to express the return value leaving the consumer to properly apply the needed
        /// engineering units for display purposes.
        /// </remarks>
        /// <returns>Calculated vars (total volt-amperes of reactive power) from imaginary and real components of specified <paramref name="voltage"/> and <paramref name="current"/> phasors.</returns>
        public static Power CalculateVars(IPhasorValue voltage, IPhasorValue current)
        {
            if (voltage == null)
                throw new ArgumentNullException(nameof(voltage), "No voltage specified");

            if (current == null)
                throw new ArgumentNullException(nameof(current), "No current specified");

            return Phasor.CalculateReactivePower(voltage.Phasor, current.Phasor);
        }

        #endregion
    }
}